sklearn/neighbors/tests/test_ball_tree.py

```diff
@@ -101,3 +101,40 @@ def one_arg_func(x):
     msg = "takes 1 positional argument but 2 were given"
     with pytest.raises(TypeError, match=msg):
         BallTree(X, metric=one_arg_func)
+
+
+@pytest.mark.parametrize("metric", itertools.chain(METRICS, BOOLEAN_METRICS))
+def test_ball_tree_numerical_consistency(metric):
+    _X = rng.random_sample((40, 3)).round(0)
+    _Y = rng.random_sample((10, 3)).round(0)
+
+    X_64 = _X.astype(dtype=np.float64)
+    Y_64 = _Y.astype(dtype=np.float64)
+
+    X_32 = _X.astype(dtype=np.float32)
+    Y_32 = _Y.astype(dtype=np.float32)
+
+    metric_params = METRICS.get(metric, {})
+    bt_64 = BallTree(X_64, leaf_size=1, metric=metric, **metric_params)
+    bt_32 = BallTree32(X_32, leaf_size=1, metric=metric, **metric_params)
+
+    # Test consistency with respect to the `query` method
+    k = 5
+    dist_64, ind_64 = bt_64.query(Y_64, k=k)
+    dist_32, ind_32 = bt_32.query(Y_32, k=k)
```

Comment:
Here, we might want to preserve dtype.

I propose asserting it.
```suggestion
    dist_64, ind_64 = bt_64.query(Y_64, k=k)
    dist_32, ind_32 = bt_32.query(Y_32, k=k)

    assert dist_64.dtype == np.float64
    assert dist_32.dtype == np.float32
```

What do you think?

Reply:
You can set a small absolute tolerance using the `atol` parameter on `assert_allclose`.

Reply:
Will that work if the ordering of the nodes found is actually changed?

Reply:
You might want to reuse or adapt this fixture:

https://github.com/scikit-learn/scikit-learn/blob/b6b6f63ebefe16403d11e8a0a2281b6e2a811933/sklearn/metrics/tests/test_pairwise_distances_reduction.py#L208-L297